syft/frameworks/torch/linalg/lr.py

```diff
@@ -284,3 +285,247 @@ def _compute_pvalues(self):
+                )
+
+            # Check if both are in the same worker
+            if not (x.child.location == c.child.location and x.child.location == y.child.location):
+                raise RuntimeError("Some tuples (X, C, y) are not located in the same worker")
+
+            # Check if they have the same size
+            x_size += x.shape[0]
+            c_size += c.shape[0]
+            y_size += y.shape[0]
+            if x_size != c_size or x_size != y_size:
+                raise ValueError("Some tuples (X, C, y) do not have the same number of samples")
+
+            # Check if all tensors have the same number of features
+            if x.shape[1] != self.n_features:
+                raise ValueError(
+                    "Transient covariate vectors do not have the same number of features"
+                )
+
+            if c.shape[1] != self.n_permanent:
+                raise ValueError(
+                    "Permanent covariate vectors do not have the same number of features"
+                )
+
+        # Set total size
+        self.total_size = x_size
+
+        # Set degrees of freedom
+        self._dgf = self.total_size - self.n_permanent - 1
+
+    @staticmethod
+    def _get_workers(ptrs):
+        """
+        Method that returns the pool of workers in a tuple
+        """
+        workers = set([])
+        for ptr in ptrs:
+            workers.add(ptr.child.location)
+        return tuple(workers)
+
+    @staticmethod
+    def _remote_dot_products(X_ptrs, C_ptrs, y_ptrs):
+        """
+        This method computes the aggregated dot-products remotely. It corresponds
+        to the Compression stage (or Compression within) of DASH algorithm
+        """
+        XX_ptrs = []
+        Xy_ptrs = []
+        yy_ptrs = []
+        CX_ptrs = []
+        Cy_ptrs = []
+        for x, c, y in zip(X_ptrs, C_ptrs, y_ptrs):
+            XX_ptrs.append(x.t() @ x)
+            Xy_ptrs.append(x.t() @ y)
+            yy_ptrs.append(y.t() @ y)
+            CX_ptrs.append(c.t() @ x)
+            Cy_ptrs.append(c.t() @ y)
+
+        return XX_ptrs, Xy_ptrs, yy_ptrs, CX_ptrs, Cy_ptrs
+
+    @staticmethod
+    def _remote_qr(C_ptrs):
+        """
+        Performs the QR decompositions of permanent covariate matrices remotely.
+        It returns a list with the upper right matrices located in each worker
+        """
+        R_ptrs = []
+        for c in C_ptrs:
+            _, r = qr(c)
+            R_ptrs.append(r)
+        return R_ptrs
+
+    @staticmethod
+    def _inv_upper(R):
+        """
+        Performs the inversion of a right upper matrix (2-dim tensor) in MPC by
+        solving the linear equation R * R_inv = I with backward substitution.
+        """
+        R_inv = torch.zeros_like(R)
+        N = R.shape[0]
+
+        # Identity Matrix
+        I = torch.zeros_like(R)
+        for i in range(N):
+            I[i, i] += 1
```

Comment:
You can use `torch.eye` I think

Reply:
The idea here is to build an identity matrix with the same workers, crypto_provider and precision fractional as `R`. 

For that, I would need whether a `torch.eye_like` (which unfortunately doesn't exist) or I am obligated to do like that to have something clean.

Reply:
👍